src/main/java/seedu/address/model/lesson/UniqueLessonList.java

```diff
@@ -0,0 +1,142 @@
+     */
+    public void add(Lesson toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            // TODO: create a DuplicateLessonException()
+            throw new DuplicatePersonException();
+        }
+        internalList.add(toAdd);
+    }
+
+    /**
+     * Replaces the lesson {@code target} in the list with {@code editedLesson}.
+     * {@code target} must exist in the list.
+     * The lesson identity of {@code editedLesson} must not be the same as another existing lesson in the list.
+     */
+    public void setLesson(Lesson target, Lesson editedLesson) {
+        requireAllNonNull(target, editedLesson);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            // TODO: create a PersonNotFoundException()
+            throw new PersonNotFoundException();
+        }
+
+        if (!target.isSameLesson(editedLesson) && contains(editedLesson)) {
+            // TODO: replace with DuplicateLessonException() once created
+            throw new DuplicatePersonException();
+        }
+
+        internalList.set(index, editedLesson);
+    }
+
+    /**
+     * Removes the equivalent lesson from the list.
+     * The lesson must exist in the list.
+     */
+    public void remove(Lesson toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
+            // TODO: replace with PersonNotFoundException() once created
+            throw new PersonNotFoundException();
+        }
+    }
+
+    public void setLessons(UniqueLessonList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
+    }
+
+    /**
+     * Replaces the contents of this list with {@code lessons}.
+     * {@code persons} must not contain duplicate persons.
+     */
+    public void setLessons(List<Lesson> lessons) {
+        requireAllNonNull(lessons);
+        if (!lessonsAreUnique(lessons)) {
+            // TODO: replace with DuplicateLessonException() once created
+            throw new DuplicatePersonException();
+        }
+
+        internalList.setAll(lessons);
+    }
+
+    /**
+     * Returns the backing list as an unmodifiable {@code ObservableList}.
+     */
+    public ObservableList<Lesson> asUnmodifiableObservableList() {
+        return internalUnmodifiableList;
+    }
+
+    @Override
+    public Iterator<Lesson> iterator() {
+        return internalList.iterator();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof UniqueLessonList // instanceof handles nulls
+                        && internalList.equals(((UniqueLessonList) other).internalList));
+    }
+
+    @Override
+    public int hashCode() {
+        return internalList.hashCode();
+    }
+
+    /**
+     * Returns true if {@code lessons} contains only unique lessons.
+     */
+    private boolean lessonsAreUnique(List<Lesson> lessons) {
```

Comment:
Perhaps you could use streams to make this cleaner, if `equals()` is the same as `isSameLesson()`.

```java
public boolean areLessonsUnique(List<Lesson> lessonList) {
    HashSet<Lesson> test = new HashSet<>();
    // HashSet add returns false if adding element that already exists
    return lessonList.stream().allMatch(x -> test.add(x)); 
}
```